Sources/DistributedActors/Plugins/ClusterSingleton/ClusterSingleton.swift

```diff
@@ -0,0 +1,448 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Distributed Actors open source project
+//
+// Copyright (c) 2019-2022 Apple Inc. and the Swift Distributed Actors project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.md for the list of Swift Distributed Actors project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Distributed
+import struct Foundation.Data
+import struct Foundation.UUID
+import Logging
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: Cluster singleton boss
+
+internal protocol ClusterSingletonBossProtocol {
+    func stop() async
+}
+
+/// Singleton wrapper of a distributed actor. The underlying singleton might run on this node, in which
+/// case `ClusterSingletonBoss` will manage its lifecycle. Otherwise, `ClusterSingletonBoss` will act as a
+/// proxy and forward calls to the remote node where the singleton is actually running. All this happens
+/// automatically and is transparent to the actor holder.
+///
+/// `ClusterSingletonBoss` has a buffer to hold remote calls temporarily in case the singleton is not available.
+/// The buffer capacity is configurable in `ClusterSingletonSettings`. Note that if the buffer becomes
+/// full, the *oldest* message would be disposed to allow insertion of the latest message.
+///
+/// `ClusterSingletonBoss` subscribes to cluster events and feeds them into `AllocationStrategy` to
+/// determine the node that the singleton runs on. If the singleton falls on *this* node, `ClusterSingletonBoss`
+/// will spawn the actual singleton actor. Otherwise, `ClusterSingletonBoss` will hand over the singleton
+/// whenever the node changes.
+internal distributed actor ClusterSingletonBoss<Act: ClusterSingletonProtocol>: ClusterSingletonBossProtocol where Act.ActorSystem == ClusterSystem {
```

Comment:
(love the boss naming, makes me smile each time lol) Enough manager object's already! 😆